v2: Complete app refactor

diogosilva30 opened the pull request:
Hi. I was exploring the repository, and I very much agree with @moritz89 regarding #12 . If there is a library [python-keycloak](https://github.com/marcospereirampj/python-keycloak) that is better maintained we should delegate the keycloak API interactions to it. 
This move heavily simplifies our code, leads to fewer possible debugs such as #38 ( which I believe it has been fixed in this version), and mainly allows us to focus our short time in integrating the package with django. From what I have seen, [python-keycloak](https://github.com/marcospereirampj/python-keycloak) seems far more stable in regards to API interaction.
Our team was facing a `jwks` error, this should be fixed aswell.

I have divided the code into 2 core and separate concepts:

`Token` --> this class is the core object of this app. If serves as a container for `access` and `refresh` tokens. Additionally, it contains several properties to retrieve information about a specific token, as well as validation mechanisms.
This class is what should be used to transfer data between the different parts of the app.

`KeycloakAdminConnector` --> this object replaces the old `Connector`, and is used to interface with Keycloak in ADMIN operations: "list users", "get user by id", etc...

I also found a lot of code that was not being used/called anywhere in the codebase, so I deleted it. A new `config.py` module with a `settings` attribute was also created to provide better parsing and single-source-of-truth settings. 

Looking forward for a detailed review and opinions.

moritz89 commented:
The general direction looks good. I have a few points, but they should all be addressable. Good points:

- Usage of python-keycloak
- Adds type hints
- Better Django settings integration
- Use of dataclasses
- Removal of dead code

Some behaviors that I would change or at least make configurable:

- Authentication without having to call Keycloak server (introspection depending on whether data is available in token)
- Include support for Async web servers (#23)
- Role/scope checks only using introspection

My main concern is the behavior of the `with_active_token_property`. It checks if the token is active, even though the default access token lifespan is only 30 minutes, which makes it largely superfluous. I'd suggest either making the check optional or using the revoked token callback functionality.

As a reference I've added the JWTs of an admin and normal user:

<details>
  <summary>Admin token</summary>

```json
{
  "exp": 1666537275,
  "iat": 1666536975,
  "auth_time": 1666536314,
  "jti": "60c114dd-f0da-4639-b345-ac17f3adac1d",
  "iss": "http://localhost:8080/auth/realms/my-realm",
  "aud": [
    "core-service",
    "account"
  ],
  "sub": "3a768e53-9a74-4504-b6e8-1570809f60b9",
  "typ": "Bearer",
  "azp": "web-app",
  "nonce": "58e0218d-75c0-43d2-804a-6340a3aa7227",
  "session_state": "19cd8ecd-f183-4738-861e-d559665f604c",
  "acr": "0",
  "allowed-origins": [
    "http://localhost:4200"
  ],
  "realm_access": {
    "roles": [
      "offline_access",
      "admin",
      "uma_authorization",
      "default-roles-my-realm"
    ]
  },
  "resource_access": {
    "core-service": {
      "roles": [
        "admin"
      ]
    },
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "openid profile email",
  "sid": "19cd8ecd-f183-4738-861e-d559665f604c",
  "email_verified": true,
  "name": "My Realm Admin",
  "preferred_username": "my-realm-admin",
  "given_name": "My Realm",
  "family_name": "Admin",
  "email": "admin@example.com"
}
```

</details>

<details>
  <summary>Normal user token</summary>

```json
{
  "exp": 1666537366,
  "iat": 1666537066,
  "auth_time": 1666537065,
  "jti": "bb3864a5-9985-456b-9709-12a6c6faf6e8",
  "iss": "http://localhost:8080/auth/realms/my-realm",
  "aud": "account",
  "sub": "be148f39-ae1b-4f80-8ee1-5d6a9fdc749c",
  "typ": "Bearer",
  "azp": "web-app",
  "nonce": "3b4f1bc8-cbc1-4293-8149-86873b5c58a5",
  "session_state": "7a1efb7d-849f-4351-be22-2038050dad76",
  "acr": "1",
  "allowed-origins": [
    "http://localhost:4200"
  ],
  "realm_access": {
    "roles": [
      "offline_access",
      "uma_authorization",
      "default-roles-my-realm"
    ]
  },
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "openid profile email",
  "sid": "7a1efb7d-849f-4351-be22-2038050dad76",
  "email_verified": true,
  "name": "Normal User",
  "preferred_username": "my-realm-user",
  "given_name": "Normal",
  "family_name": "User",
  "email": "user@example.com"
}
```

</details>

diogosilva30 commented:
@moritz89 can you provide more details for 1. and 3.? I'm kinda new to keycloak.
I think I can work on 2.

diogosilva30 commented:
@moritz89 I was checking your code. Why this block is no longer used? Is it safe to delete it?
```python
def pass_auth(self, request):
    """
    Check if the current URI path needs to skip authorization
    """
    logging.warning("Not used by middleware", DeprecationWarning, 2)
    path = request.path_info.lstrip("/")
    return any(re.match(m, path) for m in self.keycloak.exempt_uris)
```

diogosilva30 commented:
@moritz89 Please check the [changes](https://github.com/urbanplatform/django-keycloak-auth/blob/f806de702484bb45bacf024f378001b35c064f09/django_keycloak/middleware.py) to middleware regarding async capability. Also, if we push this as a major new release "2.0", what do you think of removing deprecated stuff?

moritz89 commented:
@diogosilva30 Regarding the active token check. I'd recap the OAuth flow to make sure we're on the same page.

- Authorization server --> Keycloak
- Client application --> User's browser / app
- Resource server --> Django

![image](https://user-images.githubusercontent.com/1592994/197484051-289e9ed4-3894-4c88-b3c7-f4038adf33bc.png)

The refresh token has a long lifetime (several weeks) while the access token only has a short lifetime (~3 to 120 minutes). Therefore, as the expiration time is part of the token, it is signed and short lived, it should not be necessary to check with Keycloak whether the access token is still active. By decoding the token (cryptographically signed) and checking the system time, it can be checked if the access token is valid.

The only use cases for querying Keycloak whether the access token is active is if it has been configured with a long lifetime (not recommended) or if a short (< 30 min) time to block access is required. In the second case, setting the access token life time to a few minutes is the easiest option. Otherwise it would require handling backchannel logout callbacks from Keycloak (Backchannel Logout URL option) to populate a local dict to invalidate sessions which is checked when decoding JWTs.

diogosilva30 commented:
@moritz89 I have implemented your suggestions. Please see [2bef36d](https://github.com/urbanplatform/django-keycloak-auth/pull/39/commits/2bef36db06f70fb548f4e0045027be7f843b553b). I was not aware of `exp` and `iat` properties. Token `active` property now uses these to validate instead of introspection, as in the current codebase. I've added the `TOKEN_TIMEOUT_FACTOR` to settings aswell.
I have made a few improvements that I would like to get feedback on (@moritz89 @uw-rvitorino ).

## Basic auth support
In the `KeycloakMiddleware` I added support for Basic (username+password) authorization in header. If the user passes keycloak username + password, these credentials are used to issue an `access_token`, and the `HTTP_AUTHORIZATION` is changed to `Bearer <access_token>`. 
### Why this change?
**Use case:** When working with `drf-spetacular` a custom auth schema has to be created for Keycloak. With this change a user can define the Keycloak credentials once as bellow:
<img width="648" alt="Screen Shot 2022-10-24 at 15 11 34 PM" src="https://user-images.githubusercontent.com/49190578/197546769-02895eda-36ca-416a-99b8-234fda32d8b8.png">
These credentials are then passed in each request, and our middleware takes care of converting from "Basic auth" to "Bearer token".

If you find that this behaviour should be optional, we could create a setting to enable/disable it.

moritz89 commented:
Just a note regarding the token timeout. In the previous implementation I only used a token timeout check to verify if the Django client's access token should be renewed to avoid access errors when performing admin calls to Keycloak. This was more of a hack as the proper method would have been to check the error codes, as done by [python-keycloak](https://github.com/marcospereirampj/python-keycloak/blob/83e6f54a74e2c2ad311aaaeb5618f8253126b4d4/src/keycloak/keycloak_admin.py#L3173) with `auto_refresh_token` in their `raw_get`/`raw_post` mathods.

Therefore, in this aspect the token timeout functionality is not needed any more.

In your MR you've added it for a different purpose, to check if the access token has expired. To check if a user's access token is valid (not expired, not tampered with) calling `KEYCLOAK.decode_token()` should be enough. If it has expired, it'll raise `InvalidTokenError` from `jwt.exceptions`. You can easily test this with short (e.g., 1 min lifetime access tokens).

Therefore, we can remove the explicit has expired and is active check from the MR since it is handled in `decode_token()`.

moritz89 commented:
In a different point, the audience check, this should not be disabled by default in `decode_token()`. Given two clients (web-app/public access type and core-service/confidential access type), to properly handle this case:

1. Go to clients --> web-app --> Mappers tab --> Create
2. Name --> Core service audience for web-app access tokens
3. Mapper type --> Audience (_not audience resolve_)
4. Included client audience --> core-service
5. Add to ID token --> Off
6. Add to access token --> On

For [advanced use cases](https://www.keycloak.org/docs/latest/server_admin/#_audience_resolve) (using roles and client scopes), the _audience resolve_ mapper type can be used to add the audience to the token.

An idea would be to add it as an option in settings if this should be verified.

diogosilva30 commented:
Can you implement such changes ? @moritz89

diogosilva30 commented:
@moritz89 Can you work on the mentioned changes? I'm kinda busy.

moritz89 commented:
@diogosilva30 Sorry, I had a larger commit that collided with your settings changes. I placed your 3 commits on a separate v2-complete-redo-settings-update branch. I only had a quick look, but you seemed to have added errors for importing settings?

moritz89 commented:
Also, I would not merge the changes until we have a CI with some test coverage in the changed portions as it is a significant change.

diogosilva30 commented:
@moritz89 I was having circular imports error between `config.py` and `errors.py`. I believe those are fixed now. I looked at your changes. They seemed fine, only the type annotation of Token constructor were misplaced.

moritz89 commented on django_keycloak/config.py:
```diff
@@ -0,0 +1,72 @@
+        self.KEYCLOAK_URL = f"{URL}{self.BASE_PATH}"
+
+
+__desired_variables = Settings.__annotations__.keys()
```
What are the requirements that require this complex variable extraction from the Django settings? Mainly looking at this from a maintainability perspective. The alternative is a oneliner `settings = Settings(**django_settings.KEYCLOAK_CONFIG)`

diogosilva30 replied:
I have simplified it. I do get the variables from django, but do a simple filter to clean out keys with `None` values. 
If you dont do this verification:
```
KEYCLOAK_CONFIG={
     "SERVER_URL" = os.getenv("SERVER_URL") # --> might return `None` and it is accepted by dataclass.
} 
```
If we clear out these `None` values, we can catch the `TypeError` for what variables are missing and are required. Check the updated `config.py` code :)

moritz89 replied:
Regarding the None returned when using `os.getenv()`, that's a good point

moritz89 commented on django_keycloak/managers.py:
```diff
@@ -93,31 +34,22 @@ def create_from_token(self, token, password=None, **kwargs):
-        if self.keycloak.has_superuser_perm(token):
-            is_staff = True
-            is_superuser = True
+        if token.superuser:
```
Function name should be `has_superuser_perm` or possibly be renamed to `is_superuser` here and in Token

moritz89 replied:
is_superuser would match the naming used by Django's auth module. They use it for their PermissionMixin, AbstractBaseUser and AnonymousUser, so it would match the style. There it is a boolean field, so using a property method would match that style

diogosilva30 replied:
Alright, `is_superuser` property then, let's keep the consistency

moritz89 commented on django_keycloak/config.py:
```diff
@@ -0,0 +1,72 @@
+    key: value
+    for key in __desired_variables
+    if key in __defined_variables
+    and (value := __defined_variables.get(key)) is not None
```
Should check with @uw-rvitorino and @simao-silva what the minimum target Python version should be. @simao-silva add 3.7 in the CI test matrix and the walrus operator was added in 3.8

diogosilva30 replied:
I have refactored this snippet, no more walrus operator

moritz89 commented on django_keycloak/middleware.py:
```diff
@@ -42,24 +79,28 @@ def append_user_info_to_request(self, request, token):
+        """
         auth_header = request.META.get("HTTP_AUTHORIZATION").split()
-        if len(auth_header) == 2:
+        if len(auth_header) > 1:
```
Is there a case for a valid HTTP header where it is not 2?

diogosilva30 replied:
For a strictly valid no, I guess it is redundant.

diogosilva30 replied:
You pass `Bearer mytoken x`. It will read it as `mytoken`, the old code would read it as `Bearer`. Don't know if this has any deeper consequences. What do you think?

moritz89 replied:
Cool, then we can remove `get_token()`

diogosilva30 replied:
Maybe we can add a condition to check if user correctly formatted the header?
Along the lines of: 
```python
if not (auth_type and value) or auth_type[0].lower() not in ["basic", "bearer"]:
            raise SomeError
```

moritz89 replied:
I'm for being strict with the user data in the cases that we handle, but should also not block the usage of other auth middleware (digest, hawk, aws) that also uses the auth header:

https://www.loginradius.com/blog/engineering/everything-you-want-to-know-about-authorization-headers/

diogosilva30 replied:
You are right, forgot about other middlewares 👍

diogosilva30 replied:
Also, I have removed graphene middleware. Our middleware is now a single class.

moritz89 commented on django_keycloak/middleware.py:
```diff
@@ -42,24 +79,28 @@ def append_user_info_to_request(self, request, token):
+        return auth_header[0]
 
 
 class KeycloakGrapheneMiddleware(KeycloakMiddlewareMixin):
```
Since there is a deprecation warning and a major version bump, shouldn't we remove the Graphene middleware? AFAIK it provides no additional functionality over the standard middleware but is also less efficient.

diogosilva30 replied:
I agree!

moritz89 replied:
@uw-rvitorino Is there anything from your side? Sorry, about all the requests for comment today :sweat_smile:

moritz89 commented on django_keycloak/config.py:
```diff
@@ -48,22 +48,18 @@ def __post_init__(self) -> None:
-    and (value := __defined_variables.get(key)) is not None
-}
+# Filter out configs with `None` as values
+__configs = {k: v for k, v in __configs.items() if v}
```
Doesn't this also filter out those set as False?

diogosilva30 replied:
Good catch!

moritz89 commented on django_keycloak/config.py:
```diff
@@ -48,22 +48,18 @@ def __post_init__(self) -> None:
-__desired_variables = Settings.__annotations__.keys()
-__defined_variables = getattr(
+# Get keycloak configs from django
+__configs = getattr(
```
Can this also be written as `__configs = django_settings.KEYCLOAK_CONFIG` ? It'll through an error if not defined that should be clear enough for the user

diogosilva30 replied:
Sure

moritz89 commented on django_keycloak/__init__.py:
```diff
@@ -1 +1,3 @@
+from .token import Token
```
Is there a reason for importing Token here? Would also require import update in managers.py

diogosilva30 replied:
The import here ables you to use:
```python
from django_keycloak import Token
```
Consider you are providing an "easier and public" API. Might be useful if someone decides to build a custom integration and needs to interact with the `Token` class.

moritz89 commented on django_keycloak/backends.py:
```diff
@@ -1,42 +1,71 @@
-                # Get user info based on token
-                user_info = keycloak.get_user_info(token)
+                # Get user information from token
+                user_info = token.user_info
```
We have to handle exceptions when accessing token info (JWT decoding and Keycloak API call errors). What is the preferred behavior? Letting them bubble or handling and logging them? What do other middleware apps do?

diogosilva30 replied:
Interesting question. I will have to do deeper research before providing an answer

moritz89 commented on django_keycloak/backends.py:
```diff
@@ -1,42 +1,71 @@
+            user = User.objects.create_from_token(token)
+
+        if token.is_superuser:
+            user.is_staff = user.is_superuser = True
```
This is shorter but a bit confusing. I'd rather place it on two lines OR use tuple unpacking

diogosilva30 replied:
Tbh, I think it is fine and simple

moritz89 commented on django_keycloak/errors.py:
```diff
@@ -1,3 +1,9 @@
+"""
+Module containing custom errors.
+"""
+import django_keycloak.config as config
```
This could be imported as `from django_keycloak.conf import settings`

diogosilva30 replied:
I think this was the solution to avoid the circular imports error. You are safe when importing an entire module. I will test it again

moritz89 commented on django_keycloak/middleware.py:
```diff
@@ -42,113 +82,61 @@ def append_user_info_to_request(self, request, token):
+        # 3. Request does not contain authorization header
+        # Also skip auth for "EXEMPT_URIS" defined in configs
+        if (
+            self.is_graphql_endpoint(request)
```
If we're removing special GraphQL treatment, this check can be removed

diogosilva30 replied:
Ok, consider it done

moritz89 commented on django_keycloak/mixins.py:
```diff
@@ -33,11 +33,12 @@ def tearDown(self):  # pylint: disable=invalid-name
         self.keycloak_cleanup()
```
Since the setUp and tearDown overloading has been deprecated, we might as well remove it now

diogosilva30 replied:
Why is `setUp` and `tearDown` deprecated? Did django deprecated them?

moritz89 replied:
They are not deprecated, but the method how the mixin integrated itself was a bit janky by implementing `setUp` / `tearDown` itself. It depended on the order it was included and might not call other `setUp` / `tearDown` methods.

A more robust method that also gives more flexibility is requiring the user to call `keycloak_init` and `keycloak_cleanup` in the test class `setUp` / `tearDown` methods.

diogosilva30 replied:
something like this:
```python
database = []


class MyMixin:
    def __init__(self) -> None:
        database.append([1, 2, 3])

    def __del__(self) -> None:
        while database:
            database.pop()


class MyClass(MyMixin):
    def test_something(self):
        pass


a = MyClass()

print("database: ", database) # --> database:  [[1, 2, 3]]
del a
print("database: ", database) # --> database:  []
```

moritz89 replied:
I wouldn't use del as it can be called at [any point by the garbage collector](https://stackoverflow.com/a/1481512/6783666). Also, it removes the flexibility of allowing code to be added before or after the init and cleanup methods.

moritz89 reviewed (commented):
The main aspect that is open is exception handling and getting input from the UrbanPlatform maintainers

moritz89 commented:
@diogosilva30 Do you have any changes in the pipeline? I'd push the rebased add CI tests branch once you've pushed any local changes

diogosilva30 commented:
> @diogosilva30 Do you have any changes in the pipeline? I'd push the rebased add CI tests branch once you've pushed any local changes

Tests are failing because `create_keycloak_user`. This method was removed from object manager since it was not used in the code. The sync was 1 way not 2 ways. If we need this method for tests, maybe we can create it there?

moritz89 commented:
> > @diogosilva30 Do you have any changes in the pipeline? I'd push the rebased add CI tests branch once you've pushed any local changes
> 
> Tests are failing because `create_keycloak_user`. This method was removed from object manager since it was not used in the code. The sync was 1 way not 2 ways. If we need this method for tests, maybe we can create it there?

Jupp, my app only uses it for tests. I guess placing it in the KeycloakTestMixin isn't a bad idea

diogosilva30 commented:
> > > @diogosilva30 Do you have any changes in the pipeline? I'd push the rebased add CI tests branch once you've pushed any local changes
> > 
> > 
> > Tests are failing because `create_keycloak_user`. This method was removed from object manager since it was not used in the code. The sync was 1 way not 2 ways. If we need this method for tests, maybe we can create it there?
> 
> Jupp, my app only uses it for tests. I guess placing it in the KeycloakTestMixin isn't a bad idea

After CI is green, are we ready for MR?

moritz89 commented:
> > > > @diogosilva30 Do you have any changes in the pipeline? I'd push the rebased add CI tests branch once you've pushed any local changes
> > > 
> > > 
> > > Tests are failing because `create_keycloak_user`. This method was removed from object manager since it was not used in the code. The sync was 1 way not 2 ways. If we need this method for tests, maybe we can create it there?
> > 
> > 
> > Jupp, my app only uses it for tests. I guess placing it in the KeycloakTestMixin isn't a bad idea
> 
> After CI is green, are we ready for MR?

I'll check what happens when a user tries to use an expired token. As long as that doesn't return a 500 error, from my side, it should be good enough for an initial 2.0 release. Since users should use version ranges in the requirements.txt/Pipfiles, this shouldn't break anyone's setup.

moritz89 commented on src/django_keycloak/middleware.py:
```diff
@@ -0,0 +1,143 @@
+        # 3. Request does not contain authorization header
+        # Also skip auth for "EXEMPT_URIS" defined in configs
+        if (
+            self.is_graphql_endpoint(request)
```
We should remove the GraphQL checks and references

moritz89 commented on src/django_keycloak/config.py:
```diff
@@ -0,0 +1,73 @@
+    INTERNAL_URL: Optional[str] = None
+    # Override default Keycloak base path (/auth/)
+    BASE_PATH: Optional[str] = "/auth/"
+    # Regex formatted URL to excempt the GraphQL URL"""
```
Remove GraphQL options

moritz89 commented on src/django_keycloak/middleware.py:
```diff
@@ -0,0 +1,143 @@
+
+        return any(re.match(m, path) for m in exempt_uris)
+
+    def is_graphql_endpoint(self, request):
```
Remove GraphQL related functions

moritz89 reviewed (commented):
Remove GraphQL references

moritz89 commented:
I tested v2 with my app and the admin site, GraphQL queries/mutations as well as GraphQL subscriptions work. In a few minutes I'll be able to test the behavior of expired tokens.

diogosilva30 commented:
> Remove GraphQL references

Done in [e72f410](https://github.com/urbanplatform/django-keycloak-auth/pull/39/commits/e72f410d15e0ecd5e40fb5a46e37e40af5a6034e)

moritz89 commented:
Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.

diogosilva30 commented:
> Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.

Shouldn't you get a 401?
According to djangorest [docs](https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication), if you specify  `authenticate_header` it returns a 401, if this method returns `None`, we get a 403. We did not specify it directly, but we override Django Rest [`TokenAuthentication` ](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151) which has:
```python
def authenticate_header(self, request):
        return self.keyword
```
and our `keyword` is `Bearer`. So we should get a 401 Unauthorized.

moritz89 commented:
@simao-silva You set the max-parallel jobs to 2. Is there a reason for that? Also should we add Python 3.11?

moritz89 commented:
> > Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.
> 
> Shouldn't you get a 401? According to djangorest [docs](https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication), if you specify `authenticate_header` it returns a 401, if this method returns `None`, we get a 403. We did not specify it directly, but we override Django Rest [`TokenAuthentication` ](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151) which has:
> 
> ```python
> def authenticate_header(self, request):
>         return self.keyword
> ```
> 
> and our `keyword` is `Bearer`. So we should get a 401 Unauthorized.

The 403 comes from my permission checking. Probably sees that the anonymous user has no permissions for the GraphQL query. Anyhow, Django itself will never raise 401 errors, only 403s. Reason is this: https://code.djangoproject.com/ticket/4354

diogosilva30 commented:
> > > Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.
> > 
> > 
> > Shouldn't you get a 401? According to djangorest [docs](https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication), if you specify `authenticate_header` it returns a 401, if this method returns `None`, we get a 403. We did not specify it directly, but we override Django Rest [`TokenAuthentication` ](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151) which has:
> > ```python
> > def authenticate_header(self, request):
> >         return self.keyword
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and our `keyword` is `Bearer`. So we should get a 401 Unauthorized.
> 
> The 403 comes from my permission checking. Probably sees that the anonymous user has no permissions for the GraphQL query. Anyhow, Django itself will never raise 401 errors, only 403s. Reason is this: https://code.djangoproject.com/ticket/4354

Rest framework returns 401 if you set `authenticate_header`. In your case our authenticate header is `Bearer`, so by default we must be returning 401.

moritz89 commented:
> > > > Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.
> > > 
> > > 
> > > Shouldn't you get a 401? According to djangorest [docs](https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication), if you specify `authenticate_header` it returns a 401, if this method returns `None`, we get a 403. We did not specify it directly, but we override Django Rest [`TokenAuthentication` ](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151) which has:
> > > ```python
> > > def authenticate_header(self, request):
> > >         return self.keyword
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and our `keyword` is `Bearer`. So we should get a 401 Unauthorized.
> > 
> > 
> > The 403 comes from my permission checking. Probably sees that the anonymous user has no permissions for the GraphQL query. Anyhow, Django itself will never raise 401 errors, only 403s. Reason is this: https://code.djangoproject.com/ticket/4354
> 
> Rest framework returns 401 if you set `authenticate_header`. In your case our authenticate header is `Bearer`, so by default we must be returning 401.

Just to make sure we're on the same page, if the middleware can't authenticate the user, it does not do anything to the request (only does a debug log on errors). The 4XX error is solely dependent on the Django view handling the request, so a 401 will be sent if you're using DRF and a 403 if you raise a PermissionError using Django's permission framework. So both responses are valid.

diogosilva30 commented:
> > > > > Tested a few things and looks good from my side. Server returns a nice 403 on expired tokens. If you haven't already, I can write the `create_keycloak_user` method tomorrow. After that, I have no other comments for the MR.
> > > > 
> > > > 
> > > > Shouldn't you get a 401? According to djangorest [docs](https://www.django-rest-framework.org/api-guide/authentication/#custom-authentication), if you specify `authenticate_header` it returns a 401, if this method returns `None`, we get a 403. We did not specify it directly, but we override Django Rest [`TokenAuthentication` ](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authentication.py#L151) which has:
> > > > ```python
> > > > def authenticate_header(self, request):
> > > >         return self.keyword
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > and our `keyword` is `Bearer`. So we should get a 401 Unauthorized.
> > > 
> > > 
> > > The 403 comes from my permission checking. Probably sees that the anonymous user has no permissions for the GraphQL query. Anyhow, Django itself will never raise 401 errors, only 403s. Reason is this: https://code.djangoproject.com/ticket/4354
> > 
> > 
> > Rest framework returns 401 if you set `authenticate_header`. In your case our authenticate header is `Bearer`, so by default we must be returning 401.
> 
> Just to make sure we're on the same page, if the middleware can't authenticate the user, it does not do anything to the request (only does a debug log on errors). The 4XX error is solely dependent on the Django view handling the request, so a 401 will be sent if you're using DRF and a 403 if you raise a PermissionError using Django's permission framework. So both responses are valid.

Wasn't implying your response was invalid. Sorry for misunderstand. I was just wondering if 401 was returned **when using restframework**, as their docs state. I have no problem with Django's 403.

moritz89 commented on src/django_keycloak/backends.py:
```diff
@@ -17,7 +16,7 @@ class KeycloakAuthenticationBackend(RemoteUserBackend):
         self,
         request,
-        username: Optional[str] = None,
+        remote_user: Optional[str] = None,
```
This makes it a bit confusing, as `remote_user` can now be username or Keycloak's internal user ID. We should at least document what the method expects in `remote_user`.

diogosilva30 replied:
This is due to linting warnings, due to you renaming parent method arguments name. If they pass arguments as keyword you will not catch it.

diogosilva30 replied:
But I agree with documenting

diogosilva30 replied:
Added in [c32c78c](https://github.com/urbanplatform/django-keycloak-auth/pull/39/commits/c32c78cd20e0488e04b6d7c5420742d0df3f644c)

moritz89 commented:
@diogosilva30 Jupp, lazy loading is possible, but one side-effect of the current implementation is that it'll create quite a bit of overhead since a new admin connector has to be created with each KeycloakUser object instantiation. AFAIK, the KeycloakAdmin setup is somewhat expensive with the get_token() setup.

Would an alternative be a lazy loaded global object? A bit like here: https://stackoverflow.com/a/52359211/6783666

diogosilva30 commented:
> @diogosilva30 Jupp, lazy loading is possible, but one side-effect of the current implementation is that it'll create quite a bit of overhead since a new admin connector has to be created with each KeycloakUser object instantiation. AFAIK, the KeycloakAdmin setup is somewhat expensive with the get_token() setup.
> 
> Would an alternative be a lazy loaded global object? A bit like here: https://stackoverflow.com/a/52359211/6783666

For now I would ship v2 as it is. We know it is working. But we should explore your option for optimization in further patches.

moritz89 commented:
> > @diogosilva30 Jupp, lazy loading is possible, but one side-effect of the current implementation is that it'll create quite a bit of overhead since a new admin connector has to be created with each KeycloakUser object instantiation. AFAIK, the KeycloakAdmin setup is somewhat expensive with the get_token() setup.
> > Would an alternative be a lazy loaded global object? A bit like here: https://stackoverflow.com/a/52359211/6783666
> 
> For now I would ship v2 as it is. We know it is working. But we should explore your option for optimization in further patches.

Wasn't straight forward, but I added an implementation with lazy loading. Test with `poetry run test_site/manage.py runserver` without a Keycloak server and it started without complaining. Tests also work with the Keycloak server running.

moritz89 commented on src/django_keycloak/backends.py:
```diff
@@ -0,0 +1,81 @@
+        Parameters
+        ----------
+        remote_user: str
+            The keycloak username.
```
Sorry about this comment earlier. I got confused with another piece of code. I thought `remote_user` was sometimes the username and sometimes the user_id.

I was asking myself why `get_user()` looks up `username` and then `id`. I guess its only there for compatibility reasons?

uw-rvitorino commented on src/django_keycloak/mixins.py:
```diff
@@ -0,0 +1,50 @@
+        enabled: bool = True,
+        actions: Optional[str] = None,
+    ) -> dict:
+        """Creates user on keycloak server, No state is changed on local db"""
```
keycloak --> Keycloak

uw-rvitorino commented on src/django_keycloak/token.py:
```diff
@@ -0,0 +1,252 @@
+    @ttl_cache(maxsize=1, ttl=60)
+    def public_key(self):
+        """
+        Formats the public
```
missing documentation?

uw-rvitorino commented on src/django_keycloak/token.py:
```diff
@@ -0,0 +1,252 @@
+        # Keycloak introspections return {"active": bool}
+        if "active" in info:
+            return info["active"]
+        return True
```
```python
return info["active"] if "active" in info else True
```
